File: android-beans/src/main/java/rocks/frieler/android/beans/scopes/singleton/SingletonScopedFactoryBeanHandler.java
package rocks.frieler.android.beans.scopes.singleton;

import java.util.HashMap;
import java.util.Map;

import rocks.frieler.android.beans.scopes.ScopedFactoryBean;
import rocks.frieler.android.beans.scopes.ScopedFactoryBeanHandler;

/**
 * {@link ScopedFactoryBeanHandler} for the {@value #SINGLETON_SCOPE}-scope.
 * <p>
 * It produces a new bean-instance only for the first time and re-uses it forever.
 */
public class SingletonScopedFactoryBeanHandler implements ScopedFactoryBeanHandler {
    public static final String SINGLETON_SCOPE = "singleton";

    private final Map<String, Object> beans = new HashMap<>();

    @Override
    public String getName() {
        return SINGLETON_SCOPE;
    }

    @Override
    public boolean isActive() {
        return true;
    }

    @Override
    public <T> T getBean(String name, ScopedFactoryBean<T> factoryBean) {
        @SuppressWarnings("unchecked") T bean = (T) beans.get(name);

        if (bean == null) {
            bean = factoryBean.produceBean();
            beans.put(name, bean);
        }

        return bean;
    }
}

File: android-beans/src/main/kotlin/rocks/frieler/android/beans/scopes/singleton/SingletonScopedFactoryBeanHandler.kt
package rocks.frieler.android.beans.scopes.singleton

import rocks.frieler.android.beans.BeansProvider
import rocks.frieler.android.beans.scopes.ScopedFactoryBean
import rocks.frieler.android.beans.scopes.ScopedFactoryBeanHandler
import java.util.HashMap

/**
 * [ScopedFactoryBeanHandler] for the [singleton][name]-scope.
 *
 * It produces a new bean-instance only for the first time and re-uses it forever.
 *
 * @author Christopher Frieler
 */
class SingletonScopedFactoryBeanHandler : ScopedFactoryBeanHandler {
    private val beans: MutableMap<String, Any> = HashMap()

    override val name: String
        get() = SINGLETON_SCOPE

    override val isActive: Boolean
        get() = true

    override fun <T :Any> getBean(name: String, factoryBean: ScopedFactoryBean<T>, dependencies: BeansProvider): T {
        @Suppress("UNCHECKED_CAST") var bean = beans[name] as T?
        if (bean == null) {
            bean = factoryBean.produceBean(dependencies)
            beans[name] = bean
        }
        return bean
    }

    companion object {
        const val SINGLETON_SCOPE = "singleton"
    }
}
